gusto/transport_equation.py

```diff
@@ -178,6 +186,10 @@ def advection_term(self, q):
                             dot(self.ubar('+'), self.n('+'))*q('+'))
                       + inner(self.test('-'),
                               dot(self.ubar('-'), self.n('-'))*q('-')))*self.dS
+
+        if self.outflow and self.ibp is not None:
```

Comment:
what happens if `ibp` is not None and `self.inflow` is `True`? Should this warn the user that they have set incompatible options?

Reply:
Ah have I written this a bit confusingly? This means that if `self.outflow is True` and if `ibp is not None` then we need to add the surface integrals on the edge of the domain.

And I think that if you set `ibp=None` then you get both inflow and outflow, so I think this is the only case we need to consider?

Reply:
The advection term as coded doesn't add any surface integrals if `ibp` is `None`.  So I'm confused like Jemma.

Reply:
FWIW, I note that `self.inflow` is not a thing that exists.  So is `outflow` really the right name?

Reply:
I think this would somehow be clearer if the "outflow" term were somehow added separately.  But that would necessitate much larger redesign.

The idea would follow something like what thetis does:

You have `Term` objects that implement `a` and `L`.

Then `Equation` objects don't implement the terms directly, but rather you call `Equation.add_term(TermInstance)`.

So you would have:

```python
class Term(object):
   ...

class OutflowTerm(Term):
    def residual(...):
       pass

class AdvectionEquation(Equation):
    ...

eq = AdvectionEquation(...)
if outflow:
   eq.add_term(OutflowTerm(...))
```

Or something.

But this is food for thought, rather than to be done here.

Reply:
The short answer is that this only works if `self.outflow=True` and `ibp is not None`. I'll just make this clearer by adding a warning if someone sets `ibp=None`.

What the `outflow` integral is trying to do is to add the contribution of any outgoing flux, but not incoming. This should be because `self.un` is `u` on exterior facets when `u` is pointing outwards, whilst `self.un=0` when `u` is pointing inwards.

You're right, there isn't such a thing as `self.inflow`, although we could add it in the same way if we wanted to.

With the integration by parts, I had thought that when `ibp=None`, the whole surface integral is there inside the main integral, `test*inner(u, grad(q))*dx`. When you integrate by parts, you separate out the surface integral to
`-q*div(test*u)*dx + test*q*inner(u,n)*dF`,
where `dF` is over all facets. We then separate the facets into being exterior and interior, and then normally we use boundary conditions to neglect the integral on the exterior facets. So I had presumed that if someone was choosing to use `ibp=None` then they would be trying to take include flow through the surfaces, and so wouldn't need to use the `outflow` option (which restores only outflow and not inflow).

Probably there is a mistake in something above that I said above and that's why it's confusing you!